core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1059 @@
+            } else {
+              cursor.close();
+            }
+
+            // if current value is null, combine null with next value
+            if (currentCombinedValue == null) {
+              currentCombinedValue = combineFn.apply(null, nextValueToAccumulate);
+              continue;
+            }
+
+            // if current value is "same" as next value, combine them
+            if (orderingFn.compare(currentCombinedValue, nextValueToAccumulate) == 0) {
+              currentCombinedValue = combineFn.apply(currentCombinedValue, nextValueToAccumulate);
+              continue;
+            }
+
+            // else, push accumulated value to the queue, accumulate again with next value as initial
+            outputBatch.add(currentCombinedValue);
+            batchCounter++;
+            if (batchCounter >= batchSize) {
+              outputQueue.offer(outputBatch);
+              outputBatch = new ResultBatch<>(batchSize);
+              batchCounter = 0;
+            }
+
+            // next value is now current value
+            currentCombinedValue = combineFn.apply(null, nextValueToAccumulate);
+          } else {
+            cursor.close();
+          }
+        }
+
+        if (!pQueue.isEmpty() && !cancellationGizmo.isCancelled()) {
+          // if there is still work to be done, execute a new task with the current accumulated value to continue
+          // combining where we left off
+          if (!outputBatch.isDrained()) {
+            outputQueue.offer(outputBatch);
+          }
+
+          // measure the time it took to process 'yieldAfter' elements in order to project a next 'yieldAfter' value
+          // which we want to target a 10ms task run time. smooth this value with a cumulative moving average in order
+          // to prevent normal jitter in processing time from skewing the next yield value too far in any direction
+          final long elapsedMillis = Math.max(
+              TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS),
+              1L
+          );
+          final double nextYieldAfter = Math.max(10.0 * ((double) yieldAfter / elapsedMillis), 1.0);
+          final double cumulativeMovingAverage = (nextYieldAfter + (depth * yieldAfter)) / (depth + 1);
+          final int adjustedNextYieldAfter = (int) Math.ceil(cumulativeMovingAverage);
+
+          getPool().execute(new MergeCombineAction<>(
+              pQueue,
+              outputQueue,
+              orderingFn,
+              combineFn,
+              currentCombinedValue,
+              adjustedNextYieldAfter,
+              batchSize,
+              depth + 1,
+              cancellationGizmo
+          ));
+        } else if (cancellationGizmo.isCancelled()) {
+          // if we got the cancellation signal, go ahead and write terminal value into output queue to help gracefully
+          // allow downstream stuff to stop
+          outputQueue.offer(new ResultBatch<>());
```

Comment:
Is it necessary to create a new object for the terminal value, or would it be possible to reuse a static instance?